kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Deduplicated.java

```diff
@@ -0,0 +1,236 @@
+                  if (files == null) {
+                    return Stream.empty();
+                  } else {
+                    return deduplicate(
+                            files,
+                            extractPathFromFile,
+                            (path, duplicateFiles) ->
+                                detectedDuplicateFilesInSameEventMessage(
+                                    path,
+                                    duplicateFiles,
+                                    fileType,
+                                    envelope,
+                                    tableIdentifier,
+                                    currentCommitId))
+                        .stream()
+                        .map(file -> new FileAndEnvelope<>(file, envelope));
+                  }
+                })
+            .collect(toList());
+
+    List<FileAndEnvelope<F>> result =
+        deduplicate(
+            filesAndEnvelopes,
+            fileAndEnvelope -> extractPathFromFile.apply(fileAndEnvelope.getFile()),
+            (path, duplicateFilesAndEnvelopes) ->
+                detectedDuplicateFilesAcrossMultipleEventsMessage(
+                    path, duplicateFilesAndEnvelopes, fileType, tableIdentifier, currentCommitId));
+
+    return result.stream().map(FileAndEnvelope::getFile).collect(toList());
+  }
+
+  private static class FileAndEnvelope<T> {
+    private final T file;
+    private final Envelope envelope;
+
+    FileAndEnvelope(T file, Envelope envelope) {
+      this.file = file;
+      this.envelope = envelope;
+    }
+
+    public T getFile() {
+      return file;
+    }
+
+    public Envelope getEnvelope() {
+      return envelope;
+    }
+  }
+
+  private static <T> List<T> deduplicate(
+      List<T> elements,
+      Function<T, String> keyExtractor,
+      BiFunction<String, List<T>, String> logMessageFn) {
+    Map<String, List<T>> keyToDuplicatesMapping = Maps.newHashMap();
+    elements.forEach(
+        element ->
+            keyToDuplicatesMapping
+                .computeIfAbsent(keyExtractor.apply(element), ignored -> Lists.newArrayList())
+                .add(element));
+
+    return keyToDuplicatesMapping.entrySet().stream()
+        .flatMap(
+            entry -> {
+              String key = entry.getKey();
+              List<T> values = entry.getValue();
+              if (values.size() > 1) {
+                LOG.warn(logMessageFn.apply(key, values));
+              }
+              return Stream.of(values.get(0));
+            })
+        .collect(toList());
+  }
+
+  private static <F> String detectedDuplicateFilesInSameEventMessage(
+      String path,
+      List<F> duplicateFiles,
+      String fileType,
+      Envelope envelope,
+      TableIdentifier tableIdentifier,
+      UUID currentCommitId) {
+    return String.format(
+        "Detected %d %s files with the same path=%s in the same event=%s for table=%s during commit-id=%s",
```

Comment:
Can we replace `detected` with 

`Removed` .  Yes we detected them, but we also removed them.

Tempted to put this on the `info` level as well (I know we talked about warn earlier..)

Reply:
Went with `Deduplicated` instead; `Removed` might give the wrong impression that we removed all of them (we keep one of each duplicate, remove the rest). 

If you feel you can justify `info`, let me know and we can make the change. 
`warn` feels more appropriate to me here as this should be an exceptional event but I could go either way  🤷